File: leetcode/cpp/P0046_Permutations.cpp
// https://leetcode.com/problems/permutations/

#include <iostream>
#include <vector>

class Solution {
 public:
  std::vector<std::vector<int>> permute(const std::vector<int>& nums) {
    std::vector<std::vector<int>> result;
    std::vector<int> buffer;
    backtrack(0, nums, buffer, result);
    return result;
  }

  void backtrack(int crt_idx, const std::vector<int>& nums,
                 std::vector<int>& buffer,
                 std::vector<std::vector<int>>& result) {
    if (crt_idx == nums.size()) {
      if (buffer.size() == nums.size()) {
        std::vector<int> solution;
        for (int idx : buffer) {
          solution.push_back(nums[idx]);
        }
        result.push_back(solution);
      }
    }
    for (int idx = 0; idx < nums.size(); ++idx) {
      bool found = false;
      for (int j : buffer) {
        if (j == idx) {
          found = true;
          break;
        }
      }
      if (found) {
        continue;
      }
      buffer.push_back(idx);
      backtrack(crt_idx + 1, nums, buffer, result);
      buffer.pop_back();
    }
  }
};

int main(int argc, char** argv) {
  Solution s;
  std::cout << "Test 1: " << std::endl;
  std::vector<int> nums1 = {1, 2, 3};
  std::vector<std::vector<int>> result1 = s.permute(nums1);

  for (const auto& vec : result1) {
    for (const auto& val : vec) {
      std::cout << val << " ";
    }
    std::cout << std::endl;
  }
  return 0;
}

File: leetcode/python/P0046_Permutations.py
# https://leetcode.com/problems/permutations/

from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        # typical backtracking
        # max length is 6 --> can go with n^n solution
        # i.e. not validate on every iteration

        # if crt_idx == len(nums) --> stop; if valid, add to result

        result: List[List[int]] = []

        def backtrack(crt_idx: int, buffer: List[int]):
            if crt_idx == len(nums):
                if len(set(buffer)) == len(buffer):
                    result.append([nums[idx] for idx in buffer])
                return
            for n_idx in range(len(nums)):
                # adding this and the step back below
                # increases the speed.
                if n_idx in buffer:
                    continue
                buffer[crt_idx] = n_idx
                backtrack(crt_idx + 1, buffer)
                # step back -- needed, otherwise the value will
                # stay in buffer, we generate only identity
                buffer[crt_idx] = -1

        backtrack(0, [-1 for _ in nums])
        return result


if __name__ == "__main__":
    s = Solution()
    print(f"Permutations of {[1,2,3]}: {s.permute([1, 2, 3])}")
    print(f"Permutations of {[0,1]}: {s.permute([0, 1])}")
